go/mobile/async.go

```diff
@@ -0,0 +1,214 @@
+		wg.Wait()
+		// Delete resolved tickets
+		finishedTicketsCounter := 0
+		totalTickets := len(i.Tickets.m)
+		for _, key := range finishedTickets {
+			if key != "" {
+				delete(i.Tickets.m, key)
+				finishedTicketsCounter++
+			}
+		}
+		i.Tickets.Unlock()
+		log.Info("Done checking pending tickets: ", finishedTicketsCounter, "/", totalTickets, " resolved")
+		// Stop the loop?
+		if i.Tickets.shouldStop {
+			log.Info("Stopping check pending tickets routine")
+			return
+		}
+		// Go to sleep
+		log.Info("Sleeping before checking tickets again")
+		time.Sleep(checkPendingTicketsPeriod)
+	}
+}
+
+func (t *Ticket) marshal() ([]byte, error) {
+	hdlrData, err := t.handler.marshal()
+	if err != nil {
+		return nil, err
+	}
+	t.HandlerMarshaledData = hdlrData
+	return json.Marshal(t)
+}
+
+func unmarshalTicket(data []byte) (*Ticket, error) {
+	t := &Ticket{}
+	if err := json.Unmarshal(data, t); err != nil {
+		return nil, err
+	}
+	var handler ticketInterface
+	switch t.Type {
+	case "RequestClaimStatus":
+		handler = &reqClaimStatusHandler{}
+	case "RequestClaimCredential":
+		handler = &reqClaimCredentialHandler{}
+	case "test ticket":
+		handler = &testTicketHandler{}
+	default:
+		return t, errors.New("Wrong ticket type")
+	}
+	if err := handler.unmarshal(t.HandlerMarshaledData); err != nil {
+		return t, err
+	}
+	t.handler = handler
+	return t, nil
+}
+
+func (i *Identity) loadTickets() {
+	// Load pending tickets
+	tickets := make(map[string][]byte)
+	err := db.LoadJSON(i.storage, []byte("pendingTickets"), &tickets)
+	if err == nil {
+		log.Info("Loading ", len(tickets), " pending tickets")
+		for _, t := range tickets {
+			ticket, err := unmarshalTicket(t)
+			if err != nil {
+				log.Error("Error loading ticket: ", err)
+			} else {
+				i.Tickets.m[ticket.Id] = ticket
+			}
+		}
+	} else {
+		log.Error("Error loading pending tickets: ", err)
+	}
+}
+
+func (i *Identity) storeTickets() {
+	// If the pending tickets loop is running, wait
+	i.Tickets.RLock()
+	defer i.Tickets.RUnlock()
+	// Marshal tickets
+	marshaledTickets := make(map[string][]byte)
+	log.Info("Storing ", len(i.Tickets.m), " pending tickets")
+	for _, t := range i.Tickets.m {
+		// Marshal ticket
+		data, err := t.marshal()
+		if err != nil {
+			log.Error("Error storing ticket", t.Id, err)
+		} else {
+			marshaledTickets[t.Id] = data
+		}
+	}
+	// Store pending tickets
+	tx, err := i.storage.NewTx()
+	if err != nil {
+		log.Error("Error storing ALL tickets", err)
+		return
+	}
+	if err := db.StoreJSON(tx, []byte("pendingTickets"), marshaledTickets); err != nil {
+		log.Error("Error storing ALL tickets", err)
+		return
+	}
+	if err := tx.Commit(); err != nil {
+		log.Error("Error storing ALL tickets", err)
+	}
+}
+
+// TODO: Move this code to test
+type testTicketHandler struct {
+	SayImDone bool
+	Err       string
+}
+
+func (h *testTicketHandler) isDone(id *Identity) (bool, string, error) {
+	if h.SayImDone {
+		data, err := h.marshal()
```

Comment:
Do all Tickets return themselves marshaled when `isDone` has finished?

Reply:
No, this is a special case or testing purposes